clients/src/main/java/org/apache/kafka/common/record/EosLogEntry.java

```diff
@@ -0,0 +1,392 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.record;
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.utils.AbstractIterator;
+import org.apache.kafka.common.utils.ByteBufferInputStream;
+import org.apache.kafka.common.utils.Utils;
+
+import java.io.DataInputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Iterator;
+
+import static org.apache.kafka.common.record.Records.LOG_OVERHEAD;
+
+/**
+ * LogEntry implementation for magic 2 and above. The schema is given below:
+ *
+ * LogEntry =>
+ *  BaseOffset => Int64
+ *  Length => Int32
+ *  CRC => Int32
+ *  Magic => Int8
+ *  Attributes => Int16
+ *  LastOffsetDelta => Int32
+ *  BaseTimestamp => Int64
+ *  MaxTimestamp => Int64
+ *  PID => Int64
+ *  Epoch => Int16
+ *  BaseSequence => Int32
+ *  PartitionLeaderEpoch => Int32
+ *  Records => Record1, Record2, … , RecordN
+ *
+ *  The current attributes are given below:
+ *
+ *  -----------------------------------------------------------------------------------
+ *  | Unused (5-16) | Transactional (4) | Timestamp Type (3) | Compression Type (0-2) |
+ *  -----------------------------------------------------------------------------------
```

Comment:
My guess is we'll have another message format bump before we run out of room for compression codecs, but I don't mind adding another bit if you think it's useful.